File: Codigo Fonte/src/Python/mainParallel.py
import time
import os
from parallel_merge_sort import parallel_merge_sort, LIMITE_PARALELO

RUNS = 15

# ==================== MEDICAO ====================

def medir_tempo(func, lista, profundidade):
    mem_descartada = [0]
    inicio = time.time()
    func(lista, 0, len(lista) - 1, profundidade, mem_descartada)
    return time.time() - inicio


def medir_memoria(func, lista, profundidade):
    mem = [0]
    func(lista, 0, len(lista) - 1, profundidade, mem)
    return mem[0] / 1024


# ==================== PROCESSAMENTO ====================

def processar_arquivo(caminho_arquivo):
    lista_numeros = []

    try:
        with open(caminho_arquivo, 'r') as arquivo:
            for linha in arquivo:
                linha_numeros = [int(num) for num in linha.strip().split() if num]
                if linha_numeros:
                    lista_numeros.append(linha_numeros)
    except FileNotFoundError:
        print(f"Erro: Arquivo não encontrado em {caminho_arquivo}\n")
        return

    if not lista_numeros:
        print(f"Aviso: O arquivo está vazio ou com formato inválido.\n")
        return

    tempos_parallel = []
    mems_parallel   = []

    for lista_numero in lista_numeros:
        # --- 15 runs para tempo ---
        t_parallel = 0.0
        for _ in range(RUNS):
            t_parallel += medir_tempo(parallel_merge_sort, list(lista_numero), LIMITE_PARALELO)
        tempos_parallel.append(t_parallel / RUNS)

        # --- 1 run separado para memória ---
        mems_parallel.append(medir_memoria(parallel_merge_sort, list(lista_numero), LIMITE_PARALELO))

    tamanhos = [f"10^{p}" for p in range(2, 2 + len(lista_numeros))]

    print(f"{'Tamanho':<10} {'Tempo Medio (s)':<22} {'Memoria (KB)'}")
    print("-" * 55)
    for i, tam in enumerate(tamanhos):
        print(f"{tam:<10} {tempos_parallel[i]:<22.6f} {mems_parallel[i]:.2f}")
    print()


# ==================== MAIN ====================

def main():
    base = os.path.dirname(os.path.abspath(__file__))

    arquivos_input = [
        os.path.join(base, "..", "..", "config", "input", "crescente.dat"),
        os.path.join(base, "..", "..", "config", "input", "decrescente.dat"),
        os.path.join(base, "..", "..", "config", "input", "random.dat"),
    ]

    for arquivo in arquivos_input:
        arquivo = os.path.normpath(arquivo)
        print(f"--- Processando: {arquivo} ---")
        try:
            processar_arquivo(arquivo)
        except Exception as e:
            print(f"Erro ao processar {arquivo}: {type(e).__name__}: {e}\n")


if __name__ == "__main__":
    main()

File: Codigo Fonte/src/Java/mainParallel.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;
import static java.lang.Math.pow;

public class mainParallel {
    public static void main(String[] args) throws IOException {
        List<Integer> listaNumeros, listaAux;
        ParallelMergeSort sorter = new ParallelMergeSort();

        System.out.println("Comparacoes de tempo e memoria:");

        String[] nomes    = { "Random", "Crescente", "Decrescente" };
        String[] arquivos = {
                "Codigo Fonte/config/input/random.dat",
                "Codigo Fonte/config/input/crescente.dat",
                "Codigo Fonte/config/input/decrescente.dat"
        };

        for (int f = 0; f < arquivos.length; f++) {
            listaNumeros = lerUltimaLinha.ler(arquivos[f]);
            System.out.println("\nParallel Merge Sort - " + nomes[f] + ":");
            System.out.println("Tamanho\t\tTempo Medio (s)\t\tMemoria (KB)");

            for (int pot = 2; pot <= 6; pot++) {
                int tamanho = (int) pow(10, pot);

                // --- 15 runs só para tempo ---
                long tempoTotal = 0;
                for (int i = 0; i < 15; i++) {
                    listaAux = new ArrayList<>(listaNumeros.subList(0, tamanho));
                    AtomicLong memDescartada = new AtomicLong(0);

                    long inicio = System.nanoTime();
                    sorter.paralleMergeSort(listaAux, 0, listaAux.size() - 1,
                            sorter.LIMITE_PARALELO, memDescartada);
                    long fim = System.nanoTime();
                    tempoTotal += (fim - inicio);
                }
                double tempoMedio = tempoTotal / 1_000_000_000.0 / 15;

                // --- 1 run separado só para memória ---
                listaAux = new ArrayList<>(listaNumeros.subList(0, tamanho));
                AtomicLong mem = new AtomicLong(0);
                sorter.paralleMergeSort(listaAux, 0, listaAux.size() - 1,
                        sorter.LIMITE_PARALELO, mem);
                double memoriaKb = mem.get() / 1024.0;

                System.out.printf("10^%d\t\t%.6fs\t\t%.2f KB%n", pot, tempoMedio, memoriaKb);
            }
        }
    }
}